File: cpp/src/algo_gen/robustness.cpp
/**
 * AlgoForge — src/algo_gen/robustness.cpp
 * S1 Phase 1 — Parameter robustness sweep.
 *
 * Mirrors python/algoforge/algo_gen/robustness.py logic exactly:
 *   1. Collect numeric params from indicators + risk.
 *   2. Finite-difference Sharpe gradient (quick 50-bar replay).
 *   3. Top-3 by gradient.
 *   4. 27-cell Cartesian product (±20%, 0%).
 *   5. % of cells passing the gate.
 */

#include "algo_gen_internal.hpp"

#include <algorithm>
#include <cmath>
#include <functional>
#include <map>
#include <numeric>
#include <stdexcept>
#include <string>
#include <vector>

namespace algoforge::algo_gen {

static const double PERTURB_FRAC = 0.20;
static const int    MAX_PARAMS   = 3;
static const int    QUICK_BARS   = 50;

/* =========================================================================
 * Collect numeric parameters as (path, value) pairs
 * ========================================================================= */

static std::vector<std::pair<std::string, double>>
collect_numeric_params(const AlgoManifest& m) {
    std::vector<std::pair<std::string, double>> params;

    // Indicator params
    for (const auto& ind : m.indicators) {
        for (const auto& [k, v] : ind.params) {
            params.push_back({"ind." + ind.id + "." + k, v});
        }
    }

    // Risk params
    if (m.risk.size == "atr") {
        params.push_back({"risk.atr_mult", m.risk.atr_mult});
    } else {
        params.push_back({"risk.fixed_lots", m.risk.fixed_lots});
    }
    if (m.risk.cool_down_bars > 0) {
        params.push_back({"risk.cool_down_bars", (double)m.risk.cool_down_bars});
    }

    return params;
}

/* =========================================================================
 * Apply a single parameter mutation
 * ========================================================================= */

static AlgoManifest apply_param(const AlgoManifest& m,
                                 const std::string& path,
                                 double value) {
    AlgoManifest nm = m;

    // Indicator param
    if (path.substr(0, 4) == "ind.") {
        // path = "ind.<id>.<param_name>"
        size_t p1 = 4;
        size_t p2 = path.find('.', p1);
        if (p2 == std::string::npos) return nm;
        std::string ind_id    = path.substr(p1, p2 - p1);
        std::string param_name = path.substr(p2 + 1);

        nm.indicators.clear();
        for (const auto& ind : m.indicators) {
            IndicatorDecl ni = ind;
            if (ni.id == ind_id && ni.params.count(param_name)) {
                double orig = ni.params[param_name];
                // If original was integer-like, round the perturbed value
                if (std::floor(orig) == orig) {
                    ni.params[param_name] = std::max(1.0, std::round(value));
                } else {
                    ni.params[param_name] = value;
                }
            }
            nm.indicators.push_back(ni);
        }
        return nm;
    }

    // Risk param
    if (path == "risk.atr_mult") {
        nm.risk.atr_mult = std::max(0.1, value);
    } else if (path == "risk.fixed_lots") {
        nm.risk.fixed_lots = std::max(0.001, value);
    } else if (path == "risk.cool_down_bars") {
        nm.risk.cool_down_bars = std::max(0, (int)std::round(value));
    }
    return nm;
}

/* =========================================================================
 * Public API: parameter_robustness
 * ========================================================================= */

double parameter_robustness(
    const AlgoManifest& m,
    const std::vector<Bar>& bars,
    uint64_t seed,
    const BacktestConfig& cfg
) {
    auto all_params = collect_numeric_params(m);

    if (all_params.empty()) {
        // No numeric params → perfect robustness
        return 100.0;
    }

    // Quick-replay bars
    int quick_n = std::max((int)bars.size() / 4,
                           std::min(QUICK_BARS + cfg.warmup_bars, (int)bars.size()));
    std::vector<Bar> quick_bars(bars.begin(), bars.begin() + quick_n);

    // Baseline sharpe on quick bars
    double baseline_sharpe = 0.0;
    try {
        auto r = run_manifest_backtest(m, quick_bars, cfg);
        baseline_sharpe = r.sharpe;
    } catch (...) {}

    // Finite-difference gradient for each param
    std::vector<std::tuple<double, std::string, double>> gradients;

    for (const auto& [path, val] : all_params) {
        double delta = (std::abs(val) > 1e-6) ? std::abs(val) * PERTURB_FRAC : 0.1;

        double s_plus  = baseline_sharpe;
        double s_minus = baseline_sharpe;

        try {
            auto mp = apply_param(m, path, val + delta);
            auto r  = run_manifest_backtest(mp, quick_bars, cfg);
            s_plus  = r.sharpe;
        } catch (...) {}

        try {
            auto mm = apply_param(m, path, val - delta);
            auto r  = run_manifest_backtest(mm, quick_bars, cfg);
            s_minus = r.sharpe;
        } catch (...) {}

        double grad = (delta > 1e-12) ?
            std::abs(s_plus - s_minus) / (2.0 * delta) : 0.0;
        gradients.push_back({grad, path, val});
    }

    // Sort descending by gradient
    std::sort(gradients.begin(), gradients.end(),
              [](const auto& a, const auto& b) {
                  return std::get<0>(a) > std::get<0>(b);
              });

    // Top-3
    int n_top = std::min(MAX_PARAMS, (int)gradients.size());
    std::vector<std::tuple<double, std::string, double>> top(
        gradients.begin(), gradients.begin() + n_top);

    if (top.empty()) return 100.0;

    // Cartesian product: -20%, 0%, +20% for each selected param
    std::vector<std::vector<double>> perturb_vals;
    for (const auto& [grad, path, val] : top) {
        double delta = (std::abs(val) > 1e-6) ? std::abs(val) * PERTURB_FRAC : 0.1;
        perturb_vals.push_back({val - delta, val, val + delta});
    }

    // Generate Cartesian product
    // For up to 3 params: max 27 combos
    int pass_count = 0;
    int total = 0;

    // Build index arrays for Cartesian product
    std::vector<int> idxs(n_top, 0);
    while (true) {
        // Build perturbed manifest
        AlgoManifest pm = m;
        for (int k = 0; k < n_top; ++k) {
            const auto& [grad, path, val] = top[k];
            pm = apply_param(pm, path, perturb_vals[k][idxs[k]]);
        }

        try {
            auto r = run_manifest_backtest(pm, bars, cfg);
            if (passes_gate(r)) pass_count++;
        } catch (...) {}
        total++;

        // Increment indices (Cartesian product)
        int carry = 1;
        for (int k = n_top - 1; k >= 0 && carry; --k) {
            idxs[k] += carry;
            if (idxs[k] >= 3) { idxs[k] = 0; carry = 1; }
            else               { carry = 0; }
        }
        if (carry) break;  // overflow → done
    }

    return total > 0 ? (double)pass_count / total * 100.0 : 0.0;
}

} /* namespace algoforge::algo_gen */

File: js/src/algo_gen/robustness.js
/**
 * robustness.js — Parameter robustness sweep.
 *
 * Same algorithm as Python robustness.py:
 *   1. Collect all numeric parameters.
 *   2. Finite-difference Sharpe gradient over 50-bar quick replay.
 *   3. Select top-3 most-impactful parameters.
 *   4. 27-cell Cartesian product ±20% perturbations.
 *   5. Return % of cells passing the gate.
 *
 * Must produce ±0.01 of Python's output given identical bars + manifest.
 *
 * Public API:
 *   parameterRobustness(manifest, bars, seed, cfg?) -> number (0-100)
 */

import { BTConfig, BacktestEngine } from '../backtest.js';
import { ManifestAlgo } from './runtime.js';

// ── Gate thresholds ───────────────────────────────────────────────────────────

const SHARPE_GATE   = 0.5;
const MAX_DD_GATE   = 25.0;
const MIN_TRADES    = 20;
const PERTURB_FRAC  = 0.20;
const MAX_PARAMS    = 3;
const QUICK_BARS    = 50;

// ── Internal helpers ──────────────────────────────────────────────────────────

function runBacktest(manifest, bars, cfg) {
    const algo   = new ManifestAlgo(manifest);
    const engine = new BacktestEngine(algo, cfg);
    return engine.run(bars, 'EURUSD', 3600);
}

function passesGate(result) {
    if (result.trades.length < MIN_TRADES) return false;
    const eq = result.equityCurve;
    if (eq.length === 0) return false;
    for (const v of eq) {
        if (!isFinite(v)) return false;
    }
    if (result.sharpeRatio() < SHARPE_GATE) return false;
    if (result.maxDrawdownPct() > MAX_DD_GATE) return false;
    return true;
}

/**
 * Collect all numeric parameters with their (path, value) pairs.
 * Paths mirror Python: "ind.{id}.{key}" or "risk.atr_mult" etc.
 */
function collectNumericParams(manifest) {
    const params = [];
    // Indicator params
    for (const ind of manifest.indicators) {
        for (const [k, v] of Object.entries(ind.params)) {
            if (typeof v === 'number' && typeof v !== 'boolean') {
                params.push({ path: `ind.${ind.id}.${k}`, value: v });
            }
        }
    }
    // Risk params
    const risk = manifest.risk;
    if (risk.size === 'atr') {
        params.push({ path: 'risk.atr_mult', value: risk.atrMult });
    } else {
        params.push({ path: 'risk.fixed_lots', value: risk.fixedLots });
    }
    if (risk.coolDownBars > 0) {
        params.push({ path: 'risk.cool_down_bars', value: risk.coolDownBars });
    }
    return params;
}

/**
 * Return a copy of manifest with one parameter set to a new value.
 * Mirrors Python _apply_param.
 */
function applyParam(manifest, path, newVal) {
    // Deep-copy indicators
    const newIndicators = manifest.indicators.map(ind => {
        const prefix = `ind.${ind.id}.`;
        if (path.startsWith(prefix)) {
            const paramName = path.slice(prefix.length);
            const newParams = { ...ind.params };
            const origVal   = ind.params[paramName];
            if (typeof origVal === 'number' && Number.isInteger(origVal)) {
                newParams[paramName] = Math.max(1, Math.round(newVal));
            } else {
                newParams[paramName] = newVal;
            }
            return { id: ind.id, kind: ind.kind, params: newParams };
        }
        return ind;
    });

    // Risk copy
    let risk = { ...manifest.risk };
    if (path === 'risk.atr_mult') {
        risk = { ...risk, atrMult: Math.max(0.1, newVal) };
    } else if (path === 'risk.fixed_lots') {
        risk = { ...risk, fixedLots: Math.max(0.01, newVal) };
    } else if (path === 'risk.cool_down_bars') {
        risk = { ...risk, coolDownBars: Math.max(0, Math.round(newVal)) };
    }

    return {
        ...manifest,
        indicators: newIndicators,
        risk,
    };
}

/**
 * Compute parameter robustness score (0-100).
 *
 * @param {import('./types.js').AlgoManifest} manifest
 * @param {object[]} bars
 * @param {number}   seed   (unused in robustness — kept for API parity)
 * @param {BTConfig} [cfg]
 * @returns {number}
 */
export function parameterRobustness(manifest, bars, seed, cfg) {
    if (!cfg) {
        cfg = new BTConfig({ warmupBars: Math.min(200, Math.max(20, Math.floor(bars.length / 4))) });
    }

    const allParams = collectNumericParams(manifest);

    if (allParams.length === 0) {
        // No numeric parameters — perfect robustness
        return 100.0;
    }

    // Quick-replay bars
    const quickBarsN = Math.max(
        cfg.warmupBars + 20,
        Math.min(QUICK_BARS + cfg.warmupBars, bars.length),
    );
    const quickBars = bars.slice(0, quickBarsN);

    // Baseline Sharpe
    let baseSharpe = 0;
    try {
        const baseResult = runBacktest(manifest, quickBars, cfg);
        baseSharpe = baseResult.sharpeRatio();
    } catch (_) {
        baseSharpe = 0;
    }

    // Finite-difference gradients
    const gradients = [];
    for (const { path, value } of allParams) {
        const delta = Math.abs(value) > 1e-6 ? Math.abs(value) * PERTURB_FRAC : 0.1;

        let sPlus = baseSharpe;
        try {
            const mPlus  = applyParam(manifest, path, value + delta);
            const rPlus  = runBacktest(mPlus, quickBars, cfg);
            sPlus = rPlus.sharpeRatio();
        } catch (_) {}

        let sMinus = baseSharpe;
        try {
            const mMinus = applyParam(manifest, path, value - delta);
            const rMinus = runBacktest(mMinus, quickBars, cfg);
            sMinus = rMinus.sharpeRatio();
        } catch (_) {}

        const grad = delta > 1e-12 ? Math.abs(sPlus - sMinus) / (2 * delta) : 0;
        gradients.push({ grad, path, value });
    }

    // Sort descending by gradient, pick top MAX_PARAMS
    gradients.sort((a, b) => b.grad - a.grad);
    const topParams = gradients.slice(0, MAX_PARAMS);

    if (topParams.length === 0) return 100.0;

    // Build perturbation values [-20%, 0%, +20%] for each selected param
    const perturbationValues = topParams.map(({ value }) => {
        const delta = Math.abs(value) > 1e-6 ? Math.abs(value) * PERTURB_FRAC : 0.1;
        return [value - delta, value, value + delta];
    });

    // Cartesian product over all perturbed combos
    let passCount = 0;
    let total     = 0;

    for (const combo of cartesian(perturbationValues)) {
        let perturbedManifest = manifest;
        for (let i = 0; i < topParams.length; i++) {
            perturbedManifest = applyParam(perturbedManifest, topParams[i].path, combo[i]);
        }

        try {
            const result = runBacktest(perturbedManifest, bars, cfg);
            if (passesGate(result)) passCount++;
        } catch (_) {
            // treat as failure
        }
        total++;
    }

    if (total === 0) return 0;
    return passCount / total * 100;
}

/**
 * Cartesian product of arrays.
 * @param {any[][]} arrays
 * @yields {any[]}
 */
function* cartesian(arrays) {
    if (arrays.length === 0) { yield []; return; }
    const [first, ...rest] = arrays;
    for (const v of first) {
        for (const tail of cartesian(rest)) {
            yield [v, ...tail];
        }
    }
}
